### PR TITLE
feat: add ContractDefinitionApi for Json-LD objects

### LINKED ISSUE #2835: dsp: add json-ld compliant `ContractDefinition` management api controller
# Feature Request

Same as `ContractDefinitionApi` but accepting and returning JsonLd objects 

## Which Areas Would Be Affected?
_e.g., DPF, CI, build, transfer, etc._

## Why Is the Feature Desired?
_Are there any requirements?_

## Solution Proposal
_If possible, provide a (brief!) solution proposal._

### PR DESCRIPTION
**Pull request by paullatzelsperger**
## What this PR changes/adds

Adds a "new" Contract Definition API that accepts and returns all objects in JSON-LD format.

## Why it does that

to comply with the new Dataspace Protocol

## Further notes

- transformers for `CriterionDto` <> `JsonObject` was added
- merged the `-CreateDto` and the `-UpdateDto` into the more generic `-RequestDto`, which was an abstract class before. _Existing APIs should not be affected by this._

## Linked Issue(s)

Closes #2835 

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [Etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md) for details_)

### CODE REVIEW

File: extensions/common/api/api-core/src/test/java/org/eclipse/edc/api/transformer/JsonObjectFromCriterionDtoTransformerTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    private final JsonObjectFromCriterionDtoTransformer transformer = new JsonObjectFromCriterionDtoTransformer(Json.createBuilderFactory(Map.of()), createObjectMapper());
+    private final TransformerContext context = mock(TransformerContext.class);
+
+    @Test
+    void transform() {
+        var criterion = from("foo", "=", "bar");
+        var json = transformer.transform(criterion, context);
+
+        assertThat(json).isNotNull();
+        assertThat(json.getJsonString(TYPE).getString()).isEqualTo(CRITERION_TYPE);
```
Comment (ndr-brt) [inline extensions/common/api/api-core/src/test/java/org/eclipse/edc/api/transformer/JsonObjectFromCriterionDtoTransformerTest.java:41]: assertion on values missing

File: extensions/control-plane/api/management-api/contract-definition-api/src/main/java/org/eclipse/edc/connector/api/management/contractdefinition/ContractDefinitionNewApiController.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                .map(Result::getContent)
+                .orElseThrow(() -> new ObjectNotFoundException(ContractDefinition.class, id));
+
+    }
+
+    @POST
+    @Override
+    public IdResponseDto createContractDefinition(JsonObject createObject) {
+
+        var expandResult = jsonLdService.expand(createObject);
```
Comment (ndr-brt) [inline extensions/control-plane/api/management-api/contract-definition-api/src/main/java/org/eclipse/edc/connector/api/management/contractdefinition/ContractDefinitionNewApiController.java:95]: can be composed with the transformation

File: extensions/control-plane/api/management-api/contract-definition-api/src/main/java/org/eclipse/edc/connector/api/management/contractdefinition/ContractDefinitionNewApiController.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        if (expandResult.failed()) {
+            throw new InvalidRequestException(expandResult.getFailureDetail());
```
Comment (ndr-brt) [inline extensions/control-plane/api/management-api/contract-definition-api/src/main/java/org/eclipse/edc/connector/api/management/contractdefinition/ContractDefinitionNewApiController.java:97]: the exception constructor accepts `Failure` as well

File: extensions/control-plane/api/management-api/contract-definition-api/src/main/java/org/eclipse/edc/connector/api/management/contractdefinition/ContractDefinitionNewApiController.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        var transform = transformerRegistry.transform(expandResult.getContent(), ContractDefinitionRequestDto.class)
+                .compose(dto -> transformerRegistry.transform(dto, ContractDefinition.class));
+
+        if (transform.failed()) {
```
Comment (ndr-brt) [inline extensions/control-plane/api/management-api/contract-definition-api/src/main/java/org/eclipse/edc/connector/api/management/contractdefinition/ContractDefinitionNewApiController.java:103]: `.orElseThrow`

File: system-tests/management-api/management-api-test-runner/build.gradle.kts [hunk 5]
```diff
@@ -33,6 +33,8 @@ dependencies {
     testImplementation(libs.awaitility)
     testImplementation(libs.junit.jupiter.api)
     testImplementation(libs.jakartaJson)
+    // used for the constants
+    testImplementation(project(":extensions:control-plane:api:management-api:contract-definition-api"))
```
Comment (ndr-brt) [inline system-tests/management-api/management-api-test-runner/build.gradle.kts:37]: as they are e2e tests, maybe they could not rely on constants and use the actual strings

File: extensions/common/api/api-core/src/main/java/org/eclipse/edc/api/transformer/JsonObjectFromCriterionDtoTransformer.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    public JsonObjectFromCriterionDtoTransformer(JsonBuilderFactory jsonFactory, ObjectMapper mapper) {
+        super(CriterionDto.class, JsonObject.class);
+        this.jsonFactory = jsonFactory;
+        this.mapper = mapper;
+    }
+
+    @Override
+    public @Nullable JsonObject transform(@NotNull CriterionDto criterionDto, @NotNull TransformerContext context) {
+        var builder = jsonFactory.createObjectBuilder();
+        builder.add(TYPE, EDC_NAMESPACE + "CriterionDto")
```
Comment (ronjaquensel) [inline extensions/common/api/api-core/src/main/java/org/eclipse/edc/api/transformer/JsonObjectFromCriterionDtoTransformer.java:46]: constant `CRITERION_TYPE` from `CriterionDto` can be used here

**Review by ronjaquensel — APPROVED** [review]

File: extensions/common/api/api-core/src/main/java/org/eclipse/edc/api/transformer/JsonObjectFromCriterionDtoTransformer.java (hunk 6 above)
Comment (paullatzelsperger) [reply extensions/common/api/api-core/src/main/java/org/eclipse/edc/api/transformer/JsonObjectFromCriterionDtoTransformer.java:46]: yep, must have slipped through...

**Review by ndr-brt — APPROVED** [review]

**Pull request merged**